### DISCUSSION TITLE
Understanding Grit

### DISCUSSION THREAD

**Discussion by shntnu** [broadinstitute/grit-benchmark#27] [Q&A]
@gwaygenomics 

In https://github.com/broadinstitute/grit-benchmark/blob/main/1.calculate-metrics/cell-health/0.calculate-grit.ipynb, it says

>  Some perturbations have only one guide per gene, these cannot have grit scores

But aren't you computing grit at the replicate level? And you have 6 replicates for most guides so this should be possible.

If you're computing grit at the perturbation level, are you first aggregating the profiles (to a consensus) somewhere? 

(Meta: Consider adding [Discussions](https://github.com/broadinstitute/cell-health/discussions) to this repo, and move this there?)

**Comment by gwaybio** [comment]
For this application, I am not calculating grit at the replicate level (i.e. the profile, or well, level). I am calculating grit at the perturbation level (aka the guide level). See cell 5 in that notebook you linked.

Therefore, for a single guide (which has multiple replicates) all the `r`'s from [figure 1A](https://github.com/broadinstitute/grit-benchmark/blob/main/media/Figure1A_Simple_Interpretation.png) are profiles from different guides targeting the same gene. No aggregation to consensus.

But this is a very good point, I can easily also calculate grit at the individual level. Would that help your analyses?

**Reply by shntnu** [reply]
Hm – I want to first make sure I've understood.

Let's say that a gene has k guides and n replicates per guide. 

Is the grit of a guide (as you have calculated it here) the mean across  (k-1) x n x n values? i.e. the mean across the pairwise replicate correlations of all its replicates (=n) to the replicates of all other same-gene guides (=(k-1) x n)?

**Reply by gwaybio** [reply]
Correct. This is the key code block:

https://github.com/cytomining/cytominer-eval/blob/59684cfe8a94471b45c444b38a5da0f5ede4abb0/cytominer_eval/operations/util.py#L170-L178

We look at profiles that belong to the same "group" (in the case above that means guides to the same gene) _while ignoring_ pairwise correlations to the same "replicate" (in this case replicates of the same guide perturbation).

**Reply by shntnu** [reply]
I was typing this out and didn't see your reply. I will ready your reply now. 

---

Concretely, here's what we have for a single gene from a single cell line

Each row is a well, and the value in the last column is the number of replicate correlations that can be computed for that well.

I viewed grit of AKT1-1 as two levels of summarization
1. mean (z-scored) replicate correlation per replicate, which will produce 6 values for A549:AKT-1 (because there are 6 replicates)
2. mean of the 6 values

But this is clearly wrong :D Although consistent with the figure 1A

|  id1|Metadata_cell_line |Metadata_gene_name |Metadata_pert_name | n_nonrep| n_ref| n_rep|
|----:|:------------------|:------------------|:------------------|--------:|-----:|-----:|
|  387|A549               |AKT1               |AKT1-1             |      260|    64|     5|
|  742|A549               |AKT1               |AKT1-1             |      260|    64|     5|
| 1539|A549               |AKT1               |AKT1-1             |      260|    64|     5|
| 1894|A549               |AKT1               |AKT1-1             |      260|    64|     5|
| 2307|A549               |AKT1               |AKT1-1             |      260|    64|     5|
| 2662|A549               |AKT1               |AKT1-1             |      260|    64|     5|
|  389|A549               |AKT1               |AKT1-2             |      260|    64|     5|
|  740|A549               |AKT1               |AKT1-2             |      260|    64|     5|
| 1541|A549               |AKT1               |AKT1-2             |      260|    64|     5|
| 1892|A549               |AKT1               |AKT1-2             |      260|    64|     5|
| 2309|A549               |AKT1               |AKT1-2             |      260|    64|     5|
| 2660|A549               |AKT1               |AKT1-2             |      260|    64|     5|

**Reply by shntnu** [reply]
@gwaygenomics so is this correct? (you don't need to count all the lines :D) 

Each dot is a well

![grit2](https://user-images.githubusercontent.com/1210428/108910588-133a7400-75f4-11eb-877a-b2033edbd927.jpg)

**Reply by gwaybio** [reply]
yep. That is the grit formulation in a CRISPR experiment, or when comparing compounds against MOAs. The original Figure 1A linked above is for a general case.

**Reply by shntnu** [reply]
Got it

 > We look at profiles that belong to the same "group" (in the case above that means guides to the same gene) while ignoring pairwise correlations to the same "replicate" (in this case replicates of the same guide perturbation).

If I take this literally, that would mean you would also include pairwise correlations between guide-2 and guide-3 reps (in my diagram) when computing the grit for guide-1, given that guide-2 and guide-3 belong to the same group (i.e. same gene) as guide-1, but you don't do that, correct?

**Reply by gwaybio** [reply]
Correct. We are zooming in on the specific guide's grit. We do not compare guide 2 and 3 in guide 1's grit calculation.

**Reply by shntnu** [reply]
Got it. This makes a lot of sense.

I was wondering if we could simplify the definition by computing grit on the consensus instead, but it gets messy because we do need to have several control data points for this to work well, where doing consensus won't allow for that (easily).

I'm nearly there with you. More tomorrow.

**Comment by shntnu** [comment]
@gwaygenomics Another general, non-urgent question related to Grit: Did you ever consider doing the "completely different way" described below
https://github.com/cytomining/cytominer-eval/issues/38#issuecomment-804004365. Wondering if there's a strong reason NOT to do it that way. I find the z-score approach (the current grit) more intuitive but I was wondering if there is anything particularly problematic about using AP / PRAUC.

> I'd note that there are other, completely different ways of reporting the comparison of the two distributions (1. correlation to replicates and 2. corrections to negative controls) e.g. Average Precision (replicates = class 1, negative controls = class 2), but @gwaygenomics's choice of average z-score of class 1, using class 2 as a reference, is a defensible choice.

Context: This discussion https://github.com/broadinstitute/DeepProfilerExperiments/issues/5#issuecomment-804199278 led me to ask this q.
